src/ray/common/ray_syncer/ray_syncer-inl.h

```diff
@@ -0,0 +1,160 @@
+// Copyright 2022 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+namespace ray {
+namespace syncer {
+
+/// NodeStatus keeps track of the modules in the local nodes.
+/// It contains the local components for receiving and reporting.
+/// It also keeps the raw messages receivers got.
+class NodeStatus {
+ public:
+  /// Set the local components.
+  ///
+  /// \param cid The component id.
+  /// \param reporter The reporter is defined to be the local module which wants to
+  /// broadcast its internal status to the whole clsuter. When it's null, it means there
+  /// is no reporter in this node for this component. This is the place there messages are
+  /// generated. \param receiver The receiver is defined to be the module which eventually
+  /// will have the view of of the cluster for this component. It's the place where
+  /// received messages are consumed.
+  void SetComponents(RayComponentId cid,
+                     const ReporterInterface *reporter,
+                     ReceiverInterface *receiver);
+
+  /// Get the snapshot of a component for a newer version.
+  ///
+  /// \param cid The component id to take the snapshot.
+  ///
+  /// \return If a snapshot is taken, return the message, otherwise std::nullopt.
+  std::optional<RaySyncMessage> GetSnapshot(RayComponentId cid);
+
+  /// Consume a message. Receiver will consume this message if it doesn't have
+  /// this message.
+  ///
+  /// \param message The message received.
+  ///
+  /// \return true if this node doesn't have message with newer version.
+  bool ConsumeMessage(std::shared_ptr<RaySyncMessage> message);
+
+ private:
+  /// For local nodes
+  Array<const ReporterInterface *> reporters_ = {nullptr};
+  Array<ReceiverInterface *> receivers_ = {nullptr};
+
+  /// This fields records the version of the snapshot that has been taken.
+  Array<uint64_t> snapshots_taken_ = {0};
+  /// Keep track of the latest messages received.
+  /// Use shared pointer for easier liveness management since these messages might be
+  /// sending via rpc.
+  absl::flat_hash_map<std::string, Array<std::shared_ptr<RaySyncMessage>>> cluster_view_;
```

Comment:
is storing a cluster_view_ necessary? or just storing the version is sufficient.

Reply:
storing the view will be useful for new connections which is not included in this PR yet. Given this one is already big enough, maybe next one.

Think about this case, when the cluster has no traffic and a new node added, the syncer needs to send all the information to that new added one. But unless node state change, no message will be sent, so this could be used to send the message to the newly added node.

One way to address this is to ask receiver to generate the view, but given that the message shouldn't be too big (otherwise, communication will be a very big problem), storing here is much better and make the receiver API simpler.